wcコマンドを作る

gaki273 opened the pull request.

itagakisan committed:
実行ファイルを追加

itagakisan committed:
ファイル読み込み処理を追加 機能追加

[目的]
- ファイルの内容を参照し処理を行うため

[変更内容]
- コマンドライン引数からファイルを開く
- ファイル内容をメモリに読み込む

itagakisan committed:
ファイル統計情報を処理する関数file_statisticsを定義

itagakisan committed:
file_statistics関数を実装 機能追加

[目的]
- ファイル統計計算を行う

[変更内容]
- 行数、単語数、バイト数を計算
- フォーマットした結果を返す
- メイン処理で関数を呼び出し

itagakisan committed:
標準入力から読み込みできるように処理を追加 機能追加

[目的]
- 標準入力対応を行う

[変更内容]
- 引数なしの場合は標準入力から読み込み
- パイプライン処理が可能にした

itagakisan committed:
フローズンストリングリテラルを追加

itagakisan committed:
OptionParserを使用したオプション処理を実装 機能追加

[目的]
- コマンドラインオプション解析を行う

[変更内容]
- コマンドラインオプション機能を追加

- -l: 行数のみ表示
- -w: 単語数のみ表示
- -c: 文字数のみ表示

itagakisan committed:
コマンドのオプション機能を追加 機能追加

[目的]
- OS標準のwcコマンド同様にオプションを付けることで必要な情報を出力させる

[変更内容]
- -l（行数）、-w（単語数）、-c（文字数）オプションを追加
- 複数オプションの組み合わせに対応
- 標準入力とファイル入力で出力フォーマットを調整

itagakisan committed:
カラム幅の設定を専用メソッドで行うように修正 リファクタリング

[目的]
- 重複した処理の改善、コードの可読性を改善

[変更内容]
- column_widthsメソッドを追加して幅設定を一元化
- 重複していた幅指定を削除
- コードの可読性と保守性を向上

itagakisan committed:
複数ファイル対応を追加 リファクタリング

[目的]
- 複数のファイルの引数に対応

[変更内容]
- if-else文をcase文にリファクタリング
- 引数があるかないかに加えて複数ある場合を追加

itagakisan committed:
複数ファイルを引数としたときの合計値を計算するcalc_file_statistics_totalsを追加 機能追加

[目的]
- 複数のファイルのファイル統計情報の合計値を計算

[変更内容]
- 複数ファイル処理時に合計行を追加
- 各ファイルのファイル統計情報を集計し計算

itagakisan committed:
出力結果のフォーマットformat_statisticsメソッドを追加 リファクタリング

[目的]
- 重複した処理をなくす

[変更内容]
- 行、単語、文字のフォーマットに共通のメソッドを追加
- 不要な変数、処理を削除

itagakisan committed:
オプションが1つだった場合の出力について修正

itagakisan committed:
出力順に応じたカラム幅調整に変更 修正

[目的]
- 出力のカラム幅の調整

[変更内容]
- ハッシュベースからインデックスベースの幅指定に変更
- オプションの組み合わせに関係なく出力の幅を適用

itagakisan committed:
標準入力時の単一オプションでの出力幅の設定を修正 修正

[目的]
- 標準入力の単一オプションのときの出力を修正する

[変更内容]
- ARGV.empty?内で単一オプションの判定を追加
- 標準入力でも単一オプション時は幅０を適用

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,71 @@
+
+def file_statistics(read_file, options)
+  lines = read_file.lines.count
+  words = read_file.split
```
とても細かいですが、 `lines` は1行でカウント値まで取得していて、 `words` は一端変数に配列を入れて次の行で `length` を呼んでいるのが気になりました 😅 

このような書き方の方が統一感があるかと。

```ruby
lines = read_file.lines.count
words = read_file.split.count
characters = read_file.bytesize
```

gaki273 replied:
かしこまりました！
統一感を持たせるように修正いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,71 @@
+  format_statistics(lines, words_count, characters, options).join
+end
+
+def calc_file_statistics_totals(statistics_for_each_file, options)
```
第1引数の変数名が具体的すぎるので、こちらのドキュメントを参考に抽象化できないか検討してみましょう。
https://bootcamp.fjord.jp/pages/avoid-too-specific-var-name

gaki273 replied:
かしこまりました！
ドキュメントを参考に変数名を修正いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,71 @@
+    str.strip.split.map(&:to_i)
+  end
+  sum_totals = total_statistics.transpose.map(&:sum)
+  format_statistics(sum_totals[0], sum_totals[1], sum_totals[2], options).join
```
ここは次のように書けますね。

```ruby
format_statistics(*sum_totals, options).join
```

gaki273 replied:
かしこまりました！
ご指摘のとおり、冗長で可読性が下がる書き方ですね。
修正いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,71 @@
+def format_statistics(lines, words_count, characters, options)
+  result = []
+  widths = column_widths(options)
+  result << lines if options[:lines] || options.empty?
```
毎回 `|| options.empty?` を書く必要があり条件が複雑化していますね。事前に `options.empty?` を確認することでここをシンプルにできないでしょうか？

gaki273 replied:
かしこまりました！
シンプルになるよう事前に確認できるように修正いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,71 @@
+  read_file = File.read(file)
+  puts "#{file_statistics(read_file, options)} #{file.path}"
+else
+  statistics_for_each_file = []
```
この変数名ももう少し抽象化できないか検討をお願いします 🙏

gaki273 replied:
かしこまりました！

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,71 @@
+  puts "#{file_statistics(read_file, options)} #{file.path}"
+else
+  statistics_for_each_file = []
+  ARGV.each_with_index do |filenames, index|
```
Rubyは配列に複数形の変数名をつける風習があるので `filenames` という名前からは配列を連想してしまいます。ここは配列ではないはずなので単数形を用いるようにしましょう。

gaki273 replied:
かしこまりました！
単数形に修正いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,71 @@
+    if options.values.count(true) == 1
+      [0, 0, 0]
+    else
+      [7, 8, 8]
```
words, charactersについては前の数値との区切りの半角スペースを考慮して値が1つ大きくなっている、という理解であっていますか？

もしそうであれば、全て同じ幅で `rjust` したあとに `join(' ')` した方がコードがスッキリするはずです 👍

gaki273 replied:
かしこまりました！
スッキリしたコードになるように修正いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,71 @@
+
+def calc_file_statistics_totals(statistics_for_each_file, options)
+  total_statistics = statistics_for_each_file.map do |str|
+    str.strip.split.map(&:to_i)
```
ここの `to_i` は必要でしょうか？必要な理由がありましたら教えてください 🙏

gaki273 replied:
それぞれの情報がフォーマット化され文字列として格納されていますので、複数ファイルを引き渡したとき、合計値を求める際に一旦数値に戻さないと合計値が算出されないため、`to_i`は必要であると考えます。

もし修正するとなれば、`file_statistics`メソッドの中で各情報を配列化し、複数ファイルが出力された際には合計値を返すような実装になるかと考えます。

yuuu replied:
なるほど、ここの `join` で1つの文字列になってしまっているのですね！
https://github.com/Itagaki-Shun/ruby-practices/blob/dd6dac8d59a6ae602f6dd6dabc1e2207988c25d0/05.wc/wc.rb#L20C61-L20C66

これは出力用に整形をするタイミングが速いが故に、カウントデータを用いた計算がしにくくなっている状態だと思います。整形(つまり `.join` )するタイミングを画面に出力する直前に移動させることで、ここでは `.to_i` や `.split` を呼ばずともシンプルに合計値を計算できるようにしてみましょう。

gaki273 replied:
ありがとうございます。
かしこまりました。`.join`するタイミングを見直し、計算がしやすいように修正いたします。

itagakisan committed:
変数名を変更 リファクタリング

[目的]
- コードの可読性を向上

[変更内容]
- 変数名を修正

itagakisan committed:
合計値計算の処理を修正 リファクタリング

[目的]
- 複数ファイル指定時の合計値計算処理を効率化し、コードの可読性を向上させる。

[変更内容]
- `calc_file_statistics_totals`メソッドを削除し、そのロジックを`file_statistics`メソッドに統合
- `file_statistics`メソッドにキーワード引数`is_total`を追加、合計値計算の分岐を明確化
- 単語数(`words`)の計算を統一

itagakisan committed:
`format_statistics`メソッドの条件分岐を修正 リファクタリング

[目的]
- `format_statistics`メソッドのロジックをより簡潔で可読性を高くする

[変更内容]
- if文の分岐を減らし、`select`メソッドを使用して表示する統計情報をフィルタリングするように変更
- `options`ハッシュが空の場合と、特定のオプションが指定されている場合の両方に対応

itagakisan committed:
出力フォーマットの桁揃えを行う処理を修正 リファクタリング

[目的]
- 出力フォーマットの桁揃えを、ファイルごとや合計値の最大桁数に基づいて動的に計算するように改善する

[変更内容]
- `column_widths`メソッドを`select_widths`に改名し、オプションに応じた最終的な幅を決定する処理を行う
- `calc_max_widths`メソッドを追加し、全ファイルの統計情報を事前に計算して最大桁数を算出する処理を行う
- `file_statistics`と`format_statistics`メソッドが、計算された桁幅を受け取るように変更
- メインロジックで`calc_max_widths`を先に呼び出し、その結果を各メソッドに渡すように修正

itagakisan committed:
シンボルの配列を定数に変更 リファクタリング

[目的]
- 複数のメソッドで重複して定義されていたシンボルの配列を定数として抽出し、コードの重複を削減

[変更内容]
- `KEYS`定数を定義し、`:lines`, `:words`, `:characters`のシンボル配列を格納
- 各メソッド内で配列を直接定義していた部分を、`KEYS`定数を参照するように変更

itagakisan committed:
定数のシンボル配列にfreezeを追加

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,83 @@
+  ARGV.each_with_index do |filename, index|
+    file = File.open(filename)
+    read_file = File.read(file)
+    puts "#{file_statistics(read_file, options, max_widths, statistics)} #{file.path}"
```
第4引数の `statistics` について、Rubyでは引数で渡したオブジェクトをメソッドの中で書き換える・中身の値を変えるといったことは基本避けることが多いです。呼び出し元からそのオブジェクトに書き換えが発生していることがわかりづらいためです。

メソッド中でオブジェクトを書き換えたい場合は、書き換え後のオブジェクトを戻り値として返すことが多いです。Rubyの場合は戻り値を複数指定できるのでそれを活用します 👍 

ちなみに、ここに関しては `file_statistics` の中でフォーマットをしているが故に、別途 `statistics` を返して貰う必要がでているので、フォーマットを呼び元で行うようにすれば `statistics` は渡す必要がない認識です。

gaki273 replied:
かしこまりました。
フォーマットを呼び元で行い、可読性が高くなるように修正いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,83 @@
+
+def calc_max_widths(lists, options)
+  all_statistics = []
+  lists.each do |filename|
```
ここま `map` を使ったほうが簡潔に書けますね。

gaki273 replied:
かしこまりました。
mapを使用するよう修正いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,83 @@
+  end
+end
+
+def calc_max_widths(lists, options)
```
このメソッドと `file_statistics` で同じようなカウント処理が行われていますね。処理の順番を整理することで、1つのファイルにつき、カウント処理を1回にまとめられないでしょうか？

gaki273 replied:
かしこまりました。
カウント処理を１回にまとめられるように修正いたします。

itagakisan committed:
`calc_max_widths`メソッド内のeach文をmapに変更 リファクタリング

[目的]
- 冗長な処理を無くし、コードの可読性を上げる。

[変更内容]
- `each`ループを使用した配列の組み立て処理を`map`メソッドに置き換え
- `File.open`を介さず、直接`File.read(filename)`でファイルを読み込むように修正

itagakisan committed:
統計情報の取得とフォーマット処理を分割 リファクタリング

[目的]
- 処理の責務を明確に分離し、コードの可読性と保守性を向上

[変更内容]
- `file_statistics`メソッドを、ファイル内容から統計情報（行数、単語数、バイト数）の配列を返す機能に変更
- `calc_max_widths`メソッドが、ファイル名のリストではなく統計情報のリストを引数に取るように変更
- メインロジックを再構築
  1. 全ファイルの統計情報を先に計算
  2. 最大桁幅を計算
  3. 各ファイルと合計値の統計情報をフォーマットして出力
- `format_statistics`メソッドが統計情報の配列を直接受け取るように変更

itagakisan committed:
出力する情報の桁幅を計算する処理を１つのメソッドに統合 リファクタリング

[目的]
- 桁幅を計算するロジックを単一のメソッドに集約し、コードを簡潔化

[変更内容]
- `select_widths`と`calc_max_widths`メソッドを`calc_widths`メソッドに統合
- 単一オプションが指定された際の桁幅計算が、単一ファイル入力の場合にのみ適用されるように条件を修正
